modules/hostserv/offer.c

```diff
@@ -255,8 +255,13 @@ myuser_is_in_group(struct myuser *mu, struct myentity *mt)
 static void
 take_vhost(struct sourceinfo *si, const char *host)
 {
+	char local_host[HOSTLEN+1];
 	if (strstr(host, "$account"))
-		replace(host, BUFSIZE, "$account", entity(si->smu)->name);
+	{
+		strncpy(local_host, host, HOSTLEN);
+		replace(local_host, BUFSIZE, "$account", entity(si->smu)->name);
+		host = local_host;
+	}
```

Comment:
This should use `mowgli_strlcpy()` instead of `strncpy(3)`, and the buffer should be `BUFSIZE` bytes large. At the moment it may crash because you're telling `replace()` that it's `BUFSIZE` bytes large, when it isn't. The `check_vhost_validity()` function will already do a length test.